src/Microsoft.OData.Client/ALinq/DataServiceQueryProvider.cs

```diff
@@ -172,9 +172,23 @@ internal QueryComponents Translate(Expression e)
 
             UriWriter.Translate(this.Context, addTrailingParens, e, out uri, out version);
             ResourceExpression re = e as ResourceExpression;
-            Type lastSegmentType = re.Projection == null ? re.ResourceType : re.Projection.Selector.Parameters[0].Type;
+            QueryableResourceExpression qre = e as QueryableResourceExpression;
+            Type lastSegmentType;
+
+            if (qre?.Apply != null && qre?.Apply.KeySelectorMap.Count > 0)
+            {
+                lastSegmentType = re.Projection.Selector.Parameters[1].Type.GetGenericArguments()[0];

```

Comment:
can the re or re.Projection be null?

Reply:
@habbes @ElizabethOkerio It should be null if the `KeySelectorMap` property is initialized and populated with at least one item

Reply:
@gathogojr it should be null? Did you mean it should *not* be null?

Reply:
@habbes Thanks for pointing that out. I meant "should NOT". I have fixed my comment